pywinauto/base_application.py

```diff
@@ -192,47 +197,91 @@ def __get_ctrl(self, criteria_):
             for ctrl_criteria in criteria[1:]:
                 ctrl_criteria["top_level_only"] = False
                 if "parent" not in ctrl_criteria:
                     ctrl_criteria["parent"] = previous_parent
 
                 if isinstance(ctrl_criteria["parent"], WindowSpecification):
-                    ctrl_criteria["parent"] = ctrl_criteria["parent"].find()
+                    time_left -= timestamp() - start
+                    if time_left < retry_interval:
+                        raise TimeoutError("Timed out: can not find control with given criteria {}."
+                                           "Try to increase timeout or decrease retry interval.".format(criteria[-1]))
+                    ctrl_criteria["parent"] = ctrl_criteria["parent"].find(time_left, retry_interval)
 
                 # resolve the control and return it
                 if 'backend' not in ctrl_criteria:
                     ctrl_criteria['backend'] = self.backend.name
+
                 ctrl = self.backend.generic_wrapper_class(findwindows.find_element(**ctrl_criteria))
                 previous_parent = ctrl.element_info
                 ctrls.append(ctrl)
-
-        if ctrls:
-            return (dialog, ) + tuple(ctrls)
+            return ctrls[-1]
         else:
-            return (dialog, )
+            return dialog
+
+    def __find_all_base(self, criteria_, timeout, retry_interval):
+        time_left = timeout
+        start = timestamp()
 
-    def __resolve_control(self, criteria, timeout=None, retry_interval=None):
+        if len(criteria_) == 1:
+            criteria = self._get_updated_criteria(criteria_)
+            wrapped_dialogs = []
+            dialogs = findwindows.find_elements(**criteria[0])
+            for dialog in dialogs:
+                wrapped_dialogs.append(self.backend.generic_wrapper_class(dialog))
+            return wrapped_dialogs
+
+        else:
+            time_left -= timestamp() - start
+            previous_ctrl = self.__find_base(criteria_[:-1], time_left, retry_interval)
+            previous_parent = previous_ctrl.element_info
+            ctrl_criteria = criteria_[-1]
+            ctrl_criteria["top_level_only"] = False
+            if "parent" not in ctrl_criteria:
+                ctrl_criteria["parent"] = previous_parent
+
+            if isinstance(ctrl_criteria["parent"], WindowSpecification):
+                time_left -= timestamp() - start
+                if time_left < retry_interval:
+                    raise TimeoutError("Timed out: can not find control with given criteria {}."
+                                       "Try to increase timeout or decrease retry interval.".format(ctrl_criteria))
+                ctrl_criteria["parent"] = ctrl_criteria["parent"].find(time_left, retry_interval)
+
+            # resolve the control and return it
+            if 'backend' not in ctrl_criteria:
+                ctrl_criteria['backend'] = self.backend.name
```

Comment:
Can we call `criteria = self._get_updated_criteria(criteria_)` here as we do in `__find_base`?

Reply:
Well, OK. I just saw similar code in `_get_updated_criteria` and thought it could be used. If all modifications are not necessary, it's OK.